agent/config/storage.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2019 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"sync"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Getter allows for getting a config.
+type Getter interface {
+	Get() *Config
+}
+
+// GetReloader allows for getting and reloading a config.
+type GetReloader interface {
+	Get() *Config
+	Reload(l *logrus.Entry) (string, error)
+}
+
+// Check interfaces.
+var (
+	_ Getter      = &Storage{}
+	_ GetReloader = &Storage{}
+)
+
+// NewStorage creates a new instance of Storage with optional initial config.
+func NewStorage(cfg *Config) *Storage {
+	return &Storage{
+		cfg: cfg,
+	}
+}
+
+// Storage holds config.
+type Storage struct {
+	cfg *Config
+	mu  sync.RWMutex
+}
+
+// Get returns a global config object.
+func (s *Storage) Get() *Config {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.cfg
+}
+
+// Reload reloads config into the global object.
+func (s *Storage) Reload(l *logrus.Entry) (string, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	newCfg := &Config{}
+	cfgPath, err := getFromCmdLine(newCfg, l)
+	if err != nil {
+		s.cfg = newCfg
+	}
```

Comment:
uhm...
why don't we set `s.cfg` always?
it leads us to outdated config being stored in this storage

Reply:
Good catch, did not realize there's an edge case.
Do you think checking for `ConfigFileDoesNotExistError` is sufficient here or are there any other cases where we shall reload the config regardless of errors?

Reply:
I'm not sure, I would suggest logging and returning an error and not replacing the configuration at all.

Reply:
Sorry, I see what you mean now. That's a bug, it should have returned, not replace the config.
A implemented a check for `ConfigFileDoesNotExistError` - I think this is the correct behavior since config shall be reloaded in case of this error.